### PR TITLE
[TIR] Add schedule primitive SetAxisSeparator

### PR DESCRIPTION
**Pull request by vinx13**
This PR added a schedule primitive `set_axis_separator`, it modifies `axis_separators` attribute of the target buffer, which affects the physical dimension after flattening.

cc @Lunderberg @csullivan @junrushao1994

### CODE REVIEW

File: src/tir/schedule/primitive/layout_transformation.cc [hunk 1]
```diff
@@ -17,6 +17,8 @@
  * under the License.
  */
 #include "../utils.h"
+#include "tvm/tir/schedule/block_scope.h"
+#include "tvm/tir/stmt.h"
```
Comment (junrushao1994) [inline src/tir/schedule/primitive/layout_transformation.cc:21]: nit:

```suggestion
#include <tvm/tir/schedule/block_scope.h>
#include <tvm/tir/stmt.h>
```

**Review by junrushao1994 — APPROVED** [review]

**Review by Lunderberg — APPROVED** [review]
This looks fantastic!  I really like the `ReplaceBufferMutator` utility, and look forward to using it elsewhere as well!

**Comment by tmoreau89** [comment]
cc @quic-sanirudh

**Comment by quic-sanirudh** [comment]
@Lunderberg @vinx13 This is great, thanks a lot, I would love to start using this.

I do have a small doubt however. Is there a reason why we're introducing a new schedule primitive instead of allowing the user to pass something similar to `te.AXIS_SEPARATOR` to layout_transform and internally call both `ScheduleTransformLayout` and `SetAxisSeparator` as done for te?

The reason for this question is that I discussed once with Eric, but again, the list to be passed to axis_separators argument seems like an internal detail and might confuse users. Also it might make it difficult for us to modify how axis_separators are handled internally later if needed once we expose this detail into user facing API.

File: src/tir/schedule/transform.h [hunk 2]
```diff
@@ -66,6 +72,122 @@ Array<BufferRegion> ReplaceBuffer(Array<BufferRegion> regions, const Buffer& sou
...
+      n->match_buffers = std::move(match_buffers);
+
+      Block new_block(n);
+      block_sref_reuse_->Set(GetRef<Block>(block), new_block);
+      return std::move(new_block);
+    }
+  }
+
+  /*! \brief The storage scope to be set. */
+  String storage_scope_;
```
Comment (csullivan) [inline src/tir/schedule/transform.h:181]: Looks like `storage_scope_` is leftover from refactoring StorageScopeMutator onto ReplaceBufferMutator, please delete.

File: src/tir/schedule/transform.h [hunk 3]
```diff
@@ -66,6 +72,122 @@ Array<BufferRegion> ReplaceBuffer(Array<BufferRegion> regions, const Buffer& sou
...
 
+/*!
+ * \brief A helper mutator which recursively replaces the old buffer with the new buffer and
+ * collects the block sref reuse information for the following replacement.
+ *
+ * If the buffer to be replaced in used as the source in `match_buffers`, depending the specific
+ * use cases, the target buffers in `match_buffers` may also need to be mutated. In this
+ * case, this class should be subclassed to explicitly handle `match_buffers`.
+ */
+class ReplaceBufferMutator : public StmtExprMutator {
```
Comment (csullivan) [inline src/tir/schedule/transform.h:83]: This is such a helpful visitor 💪 , thank you for introducing it.

Comment (csullivan) [reply src/tir/schedule/transform.h:83]: I can see this being helpful for some of the non-schedule TIR transformations. I don't see any src/tir/transforms including `../schedule/transform.h` but I do see some includes of `../schedule/utils.h`. Is this an appropriate place for the visitor given its potential use outside schedule transforms?

File: src/tir/schedule/transform.h [hunk 4]
```diff
@@ -66,6 +72,122 @@ Array<BufferRegion> ReplaceBuffer(Array<BufferRegion> regions, const Buffer& sou
...
+ public:
+  ReplaceBufferMutator(const Buffer& old_buffer, Buffer new_buffer,
+                       Map<Block, Block>* block_sref_reuse)
+      : block_sref_reuse_(block_sref_reuse) {
+    buffer_var_map_[old_buffer->data.get()] = std::move(new_buffer);
+  }
```
Comment (csullivan) [inline src/tir/schedule/transform.h:89]: I don't notice any other implementations defined in this header. Should we move the implementation to the source file?

**Review by csullivan — COMMENTED** [review]
Thank you for adding axis separator support to TIR scheduling! I also really love ReplaceBufferMutator, not only very useful but edifying as a concise description of all the places buffer rewrites may need to occur. Just a few comments, only one really needs addressing.

**Comment by vinx13** [comment]
@quic-sanirudh Thanks for the comments. The motivation to have a separate schedule primitive is to decouple the logical-physical mapping from generic layout transformation which doesn't have to deal with physical layout. @Lunderberg suggested that we can have a user convenience API backed by these two schedule primitives

File: src/tir/schedule/transform.h (hunk 3 above)
Comment (vinx13) [reply src/tir/schedule/transform.h:83]: `transform.h` is for transformational utilities, it is included in `../schedule/utils.h`.

File: src/tir/schedule/transform.h (hunk 4 above)
Comment (vinx13) [reply src/tir/schedule/transform.h]: sounds good, updated

**Review by csullivan — APPROVED** [review]

**Comment by quic-sanirudh** [comment]
> we can have a user convenience API backed by these two schedule primitives

Ah okay, thanks for the reply @vinx13

**Comment by csullivan** [comment]
Many thanks @vinx13 @quic-sanirudh @junrushao1994 @Lunderberg, this is merged!

**Pull request merged**